### PR TITLE
Feature/login

### PR DESCRIPTION
**Pull request by jllavec**
Login service and dashboard template added to the app with a responsive sidenav menu.
Sorry for this large pull request, but i wanted to give a dashboard template after you login.
I promise that next PR will be shorter.

@juanmct @CarlosMayo @fmarcos83

### CODE REVIEW

File: app/home/auth-factory.1.coffee [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        @showToast = showToast
+      showToast = (message) ->
+        $mdToast.show(
+          $mdToast.simple()
+            .textContent(message)
+            .position("top right")
+            .hideDelay(3000)
+        )
+      getUserData: (uid) =>
+        deferred = $q.defer()
```
Comment (fmarcos83) [inline app/home/auth-factory.1.coffee]: deferred is not used at all here

File: app/home/auth-factory.1.coffee [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+          #If something went wrong
+          @error = error
+          @message = error.message
+          @showToast(error.message)
+          return
+      login: (item) =>
+        @ref.authWithPassword(item , (error, authData) =>
+          if (error)
+            switch (error.code)
+              when "INVALID_EMAIL"
```
Comment (fmarcos83) [inline app/home/auth-factory.1.coffee]: use when "" then to avoid return

File: app/home/auth-factory.1.coffee [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        ).catch (error) ->
+          #If something went wrong
+          @error = error
+          @message = error.message
+          @showToast(error.message)
+          return
+      login: (item) =>
+        @ref.authWithPassword(item , (error, authData) =>
+          if (error)
+            switch (error.code)
```
Comment (fmarcos83) [inline app/home/auth-factory.1.coffee]: use switch error?.code to avoid switch under nested if

File: app/home/auth-factory.1.coffee [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      createUser: (item) ->
+        @auth.$createUser(
+          email: item.email
+          password: item.password).then((userData) =>
+          #If register is succesfull
+          delete item.password
+          delete item.confPass
+          @showToast('User created successfully.')
+          #Storing user data in users json
+          @ref.child("users").child(userData.uid).set(item)
```
Comment (fmarcos83) [inline app/home/auth-factory.1.coffee]: create convenience methods (setters, getters) to deal with firebase or encapsulate this into a service.

File: app/home/auth-factory.1.coffee [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+              $state.go("home.dashboard")
+            )
+          
+        )
+      logout: () =>
+        @ref.unauth()
+        @loggedIn = 0
+        @user = {}
+        @showToast('User logged out. Bye!')
+        $state.go("home.login")
```
Comment (fmarcos83) [inline app/home/auth-factory.1.coffee]: shouldn't it wait the state.og till the toast is hidden?

File: app/home/auth-factory.1.coffee [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            .position("top right")
+            .hideDelay(3000)
+        )
+      getUserData: (uid) =>
+        deferred = $q.defer()
+        request = $http.get @url
+        request.then (result) =>
+          @user=result.data[uid]
+      getUserinoData = (uid, url) ->
+        deferred = $q.defer()
```
Comment (fmarcos83) [inline app/home/auth-factory.1.coffee]: no used

File: app/home/auth-factory.1.coffee [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                @showToast("The specified user account password is incorrect.")
+                return
+              when "INVALID_USER"
+                @showToast("The specified user account does not exist.")
+                return
+              else
+                @showToast("Error logging user in:", error)
+                return
+          else
+            console.log "Authenticated successfully"
```
Comment (fmarcos83) [inline app/home/auth-factory.1.coffee]: never ever use console use $console instead

File: app/home/controllers/home-controller.coffee [hunk 8]
```diff
@@ -8,9 +8,24 @@
 
 ###
 class HomeCtrl
-  constructor: ->
+  $inject = ['AuthFactory1', '$mdSidenav']
```
Comment (fmarcos83) [inline app/home/controllers/home-controller.coffee]: change the AuthFactory1 to AuthFactory

File: app/home/templates/home.tpl.jade [hunk 9]
```diff
@@ -1,9 +1,21 @@
...
-        .md-toolbar-tools
-          span(flex="80") 
-            img(src="images/cognizant-logo.png")
-          md-button(flex="6", ui-sref-active="active", data-ui-sref='home.login') LogIn
-          md-button(flex="6", ui-sref-active="active", data-ui-sref='home.register') Register
-      md-content
-        ui-view
\ No newline at end of file
+md-toolbar#navigation
+  .md-toolbar-tools
```
Comment (fmarcos83) [inline app/home/templates/home.tpl.jade]: use ngSwitch loggedIn not logged in functionality to display the different templates acording to that state

Comment (fmarcos83) [inline app/home/templates/home.tpl.jade]: always keep in mind the less things in scope the better.
And always favor static over dynamic behaviour

File: app/home/home.less [hunk 10]
```diff
@@ -30,12 +43,20 @@
...
+  max-height:32px;
 }
 
 a.button-nav{
     height: 100%;
 }
+
+.img-circle {
+    border-radius: 50%;
+    max-width: 64px!important;
```
Comment (fmarcos83) [inline app/home/home.less]: never ever use important

File: app/home/controllers/home-controller.coffee [hunk 11]
```diff
@@ -8,9 +8,24 @@
...
+    @settings = [
+      { name: 'Dashboard', enabled: true, icon: 'fa fa-book', route: 'home.dashboard' },
+      { name: 'Exams', enabled: true, icon: 'fa fa-file-text-o', route: 'home.exams' },
+      { name: 'Stats', enabled: false, icon: 'fa fa-bar-chart', route: 'home.stats' },
+    ]
+  close: () ->
+    @$mdSidenav('left').toggle()
+  toggle: () ->
+    @$mdSidenav('left').toggle()
+  isLoggedIn: (num) ->
```
Comment (fmarcos83) [inline app/home/controllers/home-controller.coffee]: num is not very meaningfull

Comment (fmarcos83) [inline app/home/controllers/home-controller.coffee]: if you want to do this just use another method to say isNotLoggedIn or just negate isLoggedIn in the template

File: app/home/auth-factory.coffee [hunk 12]
```diff
@@ -1,43 +0,0 @@
...
-        AuthBase.ref.child("users").child(userData.uid).set(item);
-        AuthBase.message = 'User created succesfully'
-        $state.go("home.login")
-        return
-      ).catch (error) ->
-        #If something went wrong
-        AuthBase.error = error
-        AuthBase.message = error.message
-        AuthBase.showToast(error.message)
-        return
```
Comment (fmarcos83) [inline app/home/auth-factory.coffee:42]: use undefined instead of explicit return in coffeescript

File: app/home/auth-factory.1.coffee [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+        $mdToast.show(
+          $mdToast.simple()
+            .textContent(message)
+            .position("top right")
+            .hideDelay(3000)
+        )
+      getUserData: (uid) =>
+        deferred = $q.defer()
+        request = $http.get @url
+        request.then (result) =>
```
Comment (fmarcos83) [inline app/home/auth-factory.1.coffee]: why => instead of -> here?

File: app/home/auth-factory.1.coffee [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+          @user=result.data[uid]
+      getUserinoData = (uid, url) ->
```
Comment (fmarcos83) [inline app/home/auth-factory.1.coffee]: probably you mean getUserInfoData?

File: app/home/auth-factory.1.coffee [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        deferred = $q.defer()
+        request = $http.get url
+        request.then (result) ->
```
Comment (fmarcos83) [inline app/home/auth-factory.1.coffee]: I always prefer named methods in order to debug is just a sugestion. I would create onSuccessGetUserInfoData and so on...

File: e2e/home-dashboard/home-dashboard_test.coffee [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+
+describe 'Home dashboard page', ->
+  homeDashboardPage = undefined
+
+  beforeEach ->
+    homeDashboardPage = new HomeDashboardPagePo
+    browser.get '/#/dashboard'
+
+  it 'should say HomeDashboardCtrl', ->
+    expect(homeDashboardPage.heading.getText()).toEqual 'homeDashboard'
```
Comment (fmarcos83) [inline e2e/home-dashboard/home-dashboard_test.coffee:15]: this is not a good test at all texts can change

**Comment by jllavec** [comment]
Fixed all the issues you've marked, i'll merge this PR and delete branch. Thanks! :dancer:

**Pull request merged**